pkg/destroy/openstack/openstack.go

```diff
@@ -97,52 +98,63 @@ func (o *ClusterUninstaller) Run(context.Context) error {
 		"deleteFloatingIPs":    deleteFloatingIPs,
 		"deleteImages":         deleteImages,
 	}
-	returnChannel := make(chan string)
 
 	opts := &clientconfig.ClientOpts{
 		Cloud: o.Cloud,
 	}
 
+	errChan := make(chan (error), len(deleteFuncs))
+
+	var wg sync.WaitGroup
+
 	// launch goroutines
-	for name, function := range deleteFuncs {
-		go deleteRunner(name, function, opts, o.Filter, o.Logger, returnChannel)
+	for n, f := range deleteFuncs {
+		wg.Add(1)
+		go func(name string, function deleteFunc) {
+			defer wg.Done()
+			err := deleteRunner(ctx, function, opts, o.Filter, o.Logger.WithField("deleteFuncName", name))
+			if err != nil {
+				errChan <- errors.Wrapf(err, "error running %s", name)
+			}
+		}(n, f)
 	}
 
 	// wait for them to finish
-	for i := 0; i < len(deleteFuncs); i++ {
-		select {
-		case res := <-returnChannel:
-			o.Logger.Debugf("goroutine %v complete", res)
-		}
+	wg.Wait()
+
+	select {
+	case err := <-errChan:
+		return err
+	default:
+		// no errors
 	}
 
 	// we need to untag the custom network if it was provided by the user
-	err := untagRunner(opts, o.InfraID, o.Logger)
+	err := untagRunner(ctx, opts, o.InfraID, o.Logger)
 	if err != nil {
 		return err
 	}
 
 	return nil
 }
 
-func deleteRunner(deleteFuncName string, dFunction deleteFunc, opts *clientconfig.ClientOpts, filter Filter, logger logrus.FieldLogger, channel chan string) {
+func deleteRunner(ctx context.Context, dFunction deleteFunc, opts *clientconfig.ClientOpts, filter Filter, logger logrus.FieldLogger) error {
 	backoffSettings := wait.Backoff{
 		Duration: time.Second * 15,
 		Factor:   1.3,
 		Steps:    25,
 	}
 
-	err := wait.ExponentialBackoff(backoffSettings, func() (bool, error) {
+	err := exponentialBackoffWithContext(ctx, backoffSettings, func() (bool, error) {
 		return dFunction(opts, filter, logger)
 	})
 
 	if err != nil {
-		logger.Fatalf("Unrecoverable error/timed out: %v", err)
+		return errors.Wrap(err, "Unrecoverable error/timed out")
 	}
 
-	// record that the goroutine has run to completion
-	channel <- deleteFuncName
-	return
+	logger.Debugf("delete func complete")
```

Comment:
We don't need this I think, every delete function logs when exiting.

Reply:
Removing it is fine with me. It was added to replace the "goroutine {func_name} complete" debug entry that I removed.